src/librustc_typeck/check/mod.rs

```diff
@@ -4209,7 +4210,57 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         } else if let ExpectHasType(ety) = expected {
             if let Some(ref e) = blk.expr {
                 // Coerce the tail expression to the right type.
-                self.demand_coerce(e, ty, ety);
+                if let Some(mut err) = self.demand_coerce_diag(e, ty, ety) {
+                    // Be helpful when the user wrote `{... expr}` and
+                    // adding a `;` is enough to fix the error.
+                    if ety.is_nil() {
+                        let span = Span {
+                            lo: e.span.hi,
+                            hi: e.span.hi,
+                            expn_id: e.span.expn_id
+                        };
+                        err.span_label(span, &"consider adding a semicolon here");
+                    }
+
+                    // Is the block part of a fn?
+                    let parent = self.tcx.map.get(self.tcx.map.get_parent(blk.id));
+                    let fn_decl = if let Node::NodeItem(&hir::Item {
+                        name, node: hir::ItemFn(ref decl, ..), ..
+                    }) = parent {
```

Comment:
I find the AST traversal going on here to be pretty confusing. Could be improved with a comment...